Preform RBAC checks before connecting to registered OpenSSH nodes

capnspacehook opened the pull request:
Part of implementing https://github.com/gravitational/teleport/pull/19261. 

Preform RBAC checks before connecting to registered OpenSSH nodes. The RBAC logic assumes that SSH connections to registered OpenSSH nodes will always be proxied and terminated, which will come in a future PR.

capnspacehook commented on lib/srv/authhandlers.go:
```diff
@@ -361,6 +394,25 @@ func (h *AuthHandlers) UserKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*s
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		node, err = resolve.GetServer(ctx, host, port, site{
```
That's a good idea, I'll do that

capnspacehook replied:
Done

capnspacehook replied:
Oops yes, meant to do that

capnspacehook replied:
Done

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -525,6 +545,10 @@ func (h *AuthHandlers) IsHostAuthority(cert ssh.PublicKey, address string) bool
 // client) to see if this certificate can be allowed to login as user:login
 // pair to requested server.
 func (h *AuthHandlers) canLoginWithoutRBAC(cert *ssh.Certificate, clusterName string, teleportUser, osUser string) error {
+	if h.c.AuthorizeWithoutRBAC != nil {
+		return h.c.AuthorizeWithoutRBAC(cert, clusterName, teleportUser, osUser)
+	}
+
```
Can we change the abstraction here? This looks hacky in my opinion. You call `canLoginWithoutRBAC` but when `AuthorizeWithoutRBAC` is set RBAC logic won't be called at all only for testing purposes.
I think it would be cleaner to extract `canLoginWithoutRBAC` and `canLoginWithRBAC` as interfaces and provide a different implementation in the production code and tests. In case you forgot to set the code will panic which I think it's better then calling mock by mistake.

Before we go that path, what stops us from running RBAC check in tests? Sounds like something without many side effects.

capnspacehook replied:
We can run an RBAC check in the tests I added, but I don't think it would be necessary. There are already other tests that confirm that RBAC works correctly, so my thought was instead of needlessly complicating the test by creating users, roles, etc so RBAC will work, just test the a RBAC check happens or doesn't happen.

capnspacehook replied:
Done

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -87,6 +92,15 @@ type AuthHandlerConfig struct {
+	// AuthorizeWithRBAC specifies the function to preform user
+	// authorization and RBAC checks in tests. It can safely be unset
+	// otherwise.
+	AuthorizeWithRBAC func(cert *ssh.Certificate, clusterName string, node types.Server, teleportUser string, osUser string) error
```
Check my comment below.

jakule commented on lib/srv/forward/sshserver.go:
```diff
@@ -163,6 +163,8 @@ type Server struct {
 
 	targetID, targetAddr, targetHostname string
+
+	targetNode types.Server
```
Comment please

capnspacehook replied:
Done

jakule commented on lib/srv/forward/sshserver.go:
```diff
@@ -221,6 +223,8 @@ type ServerConfig struct {
 
 	TargetID, TargetAddr, TargetHostname string
+
+	TargetNode types.Server
```
Comment, please.

capnspacehook replied:
Done

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -407,9 +421,15 @@ func (h *AuthHandlers) UserKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*s
-		err = h.canLoginWithoutRBAC(cert, clusterName.GetClusterName(), teleportUser, conn.User())
+		// if the target node exists and it is an agentless node,
+		// preform an RBAC check
+		if h.c.TargetNode != nil && h.c.TargetNode.GetSubKind() == types.KindOpenSSHNode {
```
Shouldn't we return an error when `h.c.TargetNode == nil && h.c.TargetNode.GetSubKind() == types.KindOpenSSHNode` instead of allowing access without RBAC?

Also, how this case differs from the default case `h.c.Component != teleport.ComponentForwardingNode`?

capnspacehook replied:
The first case is impossible if `.c.TargetNode == nil`, did you mean `.c.TargetNode != nil && h.c.TargetNode.GetSubKind() != types.KindOpenSSHNode`?

If `h.c.TargetNode` is `nil`, the node is not registered so we can't preform a RBAC check.
If the `AuthHandlers` are running for a regular SSH node, a RBAC check should always be preformed.

jakule replied:
NVM

jakule commented on lib/srv/authhandlers_test.go:
```diff
@@ -0,0 +1,235 @@
+			component:             teleport.ComponentForwardingNode,
+			nodeExists:            true,
+			openSSHNode:           true,
+			shouldAuthWithRBAC:    true,
+			shouldAuthWithoutRBAC: false,
```
Do we need two booleans for that? Maybe
```
shouldAuthWithRBAC: [true/false]
```
is enough?

capnspacehook replied:
I want to ensure either `canLoginWithoutRBAC` or `canLoginWithRBAC` is called, and I need two booleans to do that.

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -87,6 +92,15 @@ type AuthHandlerConfig struct {
+	// AuthorizeWithRBAC specifies the function to preform user
+	// authorization and RBAC checks in tests. It can safely be unset
+	// otherwise.
+	AuthorizeWithRBAC func(cert *ssh.Certificate, clusterName string, node types.Server, teleportUser string, osUser string) error
```
Since we're embedding the interface for RBAC, why do we still need these function pointers here?

capnspacehook replied:
Oops, removed

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -521,14 +553,35 @@ func (h *AuthHandlers) IsHostAuthority(cert ssh.PublicKey, address string) bool
+}
+
+type ahLoginChecker struct {
+	h *AuthHandlers
```
Do we need `AuthHndlers` struct to just use a logger and one function? Can we reduce the requirements here?

capnspacehook replied:
Updated so only logger, config and small interface is needed, does it look good now?

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -521,14 +553,35 @@ func (h *AuthHandlers) IsHostAuthority(cert ssh.PublicKey, address string) bool
 }
 
+// loginChecker checks if the Teleport user should be able to login to
+// a node. It is created for tests.
```
This interface is extracted to enable testing, but it's not created only for testing. 
```suggestion
// a node.
```

capnspacehook replied:
Done

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -521,14 +553,35 @@ func (h *AuthHandlers) IsHostAuthority(cert ssh.PublicKey, address string) bool
-func (h *AuthHandlers) canLoginWithoutRBAC(cert *ssh.Certificate, clusterName string, teleportUser, osUser string) error {
-	h.log.Debugf("Checking permissions for (%v,%v) to login to node without RBAC checks.", teleportUser, osUser)
+func (a *ahLoginChecker) canLoginWithoutRBAC(cert *ssh.Certificate, clusterName string, teleportUser, osUser string) error {
+	if a.h.c.AuthorizeWithoutRBAC != nil {
+		return a.h.c.AuthorizeWithoutRBAC(cert, clusterName, teleportUser, osUser)
+	}
```
As we talked about earlier, please remove this redirection. It's not needed anymore since the `AuthHandler` can accept different implementations of `loginChecker` interface.

capnspacehook replied:
Done

capnspacehook commented:
@rosstimothy can I have another review please?

capnspacehook commented on lib/srv/authhandlers.go:
```diff
@@ -521,14 +544,41 @@ func (h *AuthHandlers) IsHostAuthority(cert ssh.PublicKey, address string) bool
+
+// authorityRetriever returns a CA of type caType if key is signed by a
+// CA of type caType.
+type authorityRetriever interface {
+	// authorityForCert checks if the certificate was signed by a Teleport
+	// Certificate Authority and returns it.
+	authorityForCert(caType types.CertAuthType, key ssh.PublicKey) (types.CertAuthority, error)
+}
```
I guess that's true. I also thought `canLoginWithoutRBAC` was called multiple times, but it's only called once. So removing it shouldn't be a big deal.

capnspacehook replied:
Done

capnspacehook commented on lib/srv/authhandlers.go:
```diff
@@ -79,6 +79,10 @@ type AuthHandlerConfig struct {
 
+	// TargetNode is details about the node we are trying to connect to.
+	// It will only be set when the node is an agentless or teleport node.
+	TargetNode types.Server
```
Done

capnspacehook commented on lib/srv/authhandlers.go:
```diff
@@ -521,14 +544,41 @@ func (h *AuthHandlers) IsHostAuthority(cert ssh.PublicKey, address string) bool
+	// canLoginWithRBAC checks the given certificate (supplied by a connected
+	// client) to see if this certificate can be allowed to login as user:login
+	// pair to requested server and if RBAC rules allow login.
+	canLoginWithRBAC(cert *ssh.Certificate, clusterName string, node types.Server, teleportUser, osUser string) error
```
Done

capnspacehook commented on lib/srv/authhandlers.go:
```diff
@@ -539,24 +589,24 @@ func (h *AuthHandlers) canLoginWithoutRBAC(cert *ssh.Certificate, clusterName st
 // client) to see if this certificate can be allowed to login as user:login
 // pair to requested server and if RBAC rules allow login.
-func (h *AuthHandlers) canLoginWithRBAC(cert *ssh.Certificate, clusterName string, teleportUser, osUser string) error {
+func (a *ahLoginChecker) canLoginWithRBAC(cert *ssh.Certificate, clusterName string, node types.Server, teleportUser, osUser string) error {
```
Done

capnspacehook commented on lib/srv/authhandlers_test.go:
```diff
@@ -0,0 +1,235 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// create node resource
+			var node types.Server
```
Done

capnspacehook commented on lib/srv/authhandlers_test.go:
```diff
@@ -0,0 +1,235 @@
+		nodeExists            bool
+		openSSHNode           bool
+		shouldAuthWithRBAC    bool
+		shouldAuthWithoutRBAC bool
```
Done

capnspacehook commented on lib/srv/forward/sshserver.go:
```diff
@@ -163,6 +163,10 @@ type Server struct {
+
+	// targetServer is information about the target node.
+	// It will only be set when the node is an agentless or teleport node.
+	targetNode types.Server
```
Done

capnspacehook commented on lib/srv/forward/sshserver.go:
```diff
@@ -221,6 +225,9 @@ type ServerConfig struct {
 	TargetID, TargetAddr, TargetHostname string
+
+	// TargetNode is information about the target node.
+	TargetNode types.Server
```
Done

capnspacehook commented on lib/srv/authhandlers_test.go:
```diff
@@ -0,0 +1,235 @@
+				n, err := types.NewServer("testie_node", types.KindNode, types.ServerSpecV2{
+					Addr:     "1.2.3.4:22",
+					Hostname: "testie",
+					Version:  "v2",
```
Done

jakule commented on lib/srv/authhandlers.go:
```diff
@@ -404,12 +417,25 @@ func (h *AuthHandlers) UserKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*s
 		return permissions, nil
 	}
 
+	ca, err := h.authorityForCert(types.UserCA, cert.SignatureKey)
```
@capnspacehook Can you please add a comment that this check is needed even if we don't need the full RBAC to check to make it less likely that someone will remove it in the future?

capnspacehook replied:
Good call, done

jakule reviewed (approved):
LGTM with a few nits

capnspacehook commented on lib/srv/authhandlers.go:
```diff
@@ -404,12 +417,25 @@ func (h *AuthHandlers) UserKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*s
 
+	ca, err := h.authorityForCert(types.UserCA, cert.SignatureKey)
+	if err != nil {
+		log.Errorf("Permission denied: %v", err)
```
I did that to keep logging behavior the same as before. Calling `authorityForCert` is essentially all that `canLoginWithoutRBAC` did, and if it failed the error would be logged in the same way below.

capnspacehook commented on lib/srv/authhandlers_test.go:
```diff
@@ -0,0 +1,223 @@
+	}
+}
+
+func TestRBAC(t *testing.T) {
```
Good catch, done

rosstimothy reviewed (commented):
LGTM thanks for addressing all of my comments. I'm going to hold off on approving until the RFD has been approved and merged though.

capnspacehook commented:
@rosstimothy RFD is in the merge queue currently, so this should be good to merge now

capnspacehook committed:
preform RBAC checks before connecting to registered OpenSSH nodes

The RBAC logic assumes that SSH connections to registered OpenSSH nodes
will always be proxied and terminated, which will come in a future PR.

Part of implementing #19261.
